### PR TITLE
feat(rdpdr): `DR_CORE_SERVER_CLIENTID_CONFIRM` and `DR_CORE_DEVICELIST_ANNOUNCE`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds handling for `DR_CORE_SERVER_CLIENTID_CONFIRM` and `DR_CORE_DEVICELIST_ANNOUNCE`/`DR_CORE_DEVICELIST_ANNOUNCE_REQ`, the next steps in the rdpdr initialization sequence.

### CODE REVIEW

**Review by CBenoit — COMMENTED** [review]
Looking quite good to me.
Small request: can you add your PDU structures to the [`pdu_decode` fuzzing oracle](https://github.com/Devolutions/IronRDP/blob/babbd68af576f87044d99b2c091d3e0cf934d835/crates/ironrdp-fuzzing/src/oracles/mod.rs#L15-L78)?

File: crates/ironrdp-rdpdr/src/pdu/efs.rs [hunk 1]
```diff
@@ -263,6 +267,53 @@ impl CoreCapabilityKind {
     }
 }
 
+#[derive(Debug)]
+pub struct Capabilities(Vec<CapabilityMessage>);
+
+impl Capabilities {
+    pub fn new() -> Self {
+        let mut this = Self(vec![]);
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/pdu/efs.rs]: I changed them all for consistency's sake: d429a94c76e81aca8ed19c71e32af153971b28c7

File: crates/ironrdp-rdpdr/src/lib.rs [hunk 2]
```diff
@@ -27,25 +28,37 @@ use tracing::{trace, warn};
...
+    /// All devices not of the type [`DeviceType::Filesystem`] must be declared here.
+    device_list: Devices,
 }
 
 impl Default for Rdpdr {
     fn default() -> Self {
-        Self::new("IronRDP".to_string(), vec![CapabilityMessage::new_general(0)])
+        Self::new("IronRDP".to_string())
     }
 }
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply crates/ironrdp-rdpdr/src/lib.rs]: https://github.com/Devolutions/IronRDP/pull/193/commits/4660c38ce604514ec289f7f3f72e37988cef5c17

**Comment by ibeckermayer** [comment]
> Looking quite good to me. Small request: can you add your PDU structures to the [`pdu_decode` fuzzing oracle](https://github.com/Devolutions/IronRDP/blob/babbd68af576f87044d99b2c091d3e0cf934d835/crates/ironrdp-fuzzing/src/oracles/mod.rs#L15-L78)?

b9b2ba92c2d5e760a5cf29b4c6b6e9392eadc778

**Review by CBenoit — APPROVED** [review]
Thank you, looks good to me!

**Review by pacmancoder — APPROVED** [review]
LGTM

**Comment by CBenoit** [comment]
> Unfortunately these structs rely on taking ownership of the underlying
data when it's used. This would work very nicely/efficiently, except
under normal circumstances ironrdp-client connects and then reconnects
during a session, which means that in the second run (the one that the user
actually interacts with) the data is gone, and empty lists are sent for
both of these values.

This is very surprising by the way. Currently ironrdp-client is reconnecting from scratch and the Rdpdr channel is constructed again from scratch, so `with_smartcard` is called again and the data should be created again. Why would it be empty?

**Comment by ibeckermayer** [comment]
> > Unfortunately these structs rely on taking ownership of the underlying
> > data when it's used. This would work very nicely/efficiently, except
> > under normal circumstances ironrdp-client connects and then reconnects
> > during a session, which means that in the second run (the one that the user
> > actually interacts with) the data is gone, and empty lists are sent for
> > both of these values.
> 
> This is very surprising by the way. Currently ironrdp-client is reconnecting from scratch and the Rdpdr channel is thus constructed from scratch as well, meaning `with_smartcard` is called again. I would expect the data to be populated again. Why are we actually ending up with empty lists?

@CBenoit you're correct, I was misinterpreting the logs. What I'm seeing on that commit is that we're going through the RDPDR initialization sequence twice in a single connection:

```
2023-09-20T21:49:23.076710Z TRACE ironrdp_rdpdr: received RdpdrPdu(VersionAndIdPdu { version_major: 1, version_minor: 13, client_id: 3, kind: ServerAnnounceRequest })
2023-09-20T21:49:23.076720Z TRACE ironrdp_rdpdr: sending RdpdrPdu(VersionAndIdPdu { version_major: 1, version_minor: 12, client_id: 3, kind: ClientAnnounceReply })
2023-09-20T21:49:23.076727Z TRACE ironrdp_rdpdr: sending RdpdrPdu(Unicode("IronRDP"))
2023-09-20T21:49:23.308538Z TRACE ironrdp_rdpdr: received RdpdrPdu(CoreCapability { padding: 0, capabilities: [CapabilityMessage { header: CapabilityHeader { cap_type: General, length: 44, version: 2 }, capability_data: General(GeneralCapabilitySet { os_type: 2, os_version: 0, protocol_major_version: 1, protocol_minor_version: 13, io_code_1: IoCode1(RDPDR_IRP_MJ_CREATE | RDPDR_IRP_MJ_CLEANUP | RDPDR_IRP_MJ_CLOSE | RDPDR_IRP_MJ_READ | RDPDR_IRP_MJ_WRITE | RDPDR_IRP_MJ_FLUSH_BUFFERS | RDPDR_IRP_MJ_SHUTDOWN | RDPDR_IRP_MJ_DEVICE_CONTROL | RDPDR_IRP_MJ_QUERY_VOLUME_INFORMATION | RDPDR_IRP_MJ_SET_VOLUME_INFORMATION | RDPDR_IRP_MJ_QUERY_INFORMATION | RDPDR_IRP_MJ_SET_INFORMATION | RDPDR_IRP_MJ_DIRECTORY_CONTROL | RDPDR_IRP_MJ_LOCK_CONTROL | RDPDR_IRP_MJ_QUERY_SECURITY | RDPDR_IRP_MJ_SET_SECURITY), io_code_2: 0, extended_pdu: ExtendedPdu(RDPDR_DEVICE_REMOVE_PDUS | RDPDR_CLIENT_DISPLAY_NAME_PDU | RDPDR_USER_LOGGEDON_PDU), extra_flags_1: ExtraFlags1(0x0), extra_flags_2: 0, special_type_device_cap: 2 }) }, CapabilityMessage { header: CapabilityHeader { cap_type: Printer, length: 8, version: 1 }, capability_data: Printer }, CapabilityMessage { header: CapabilityHeader { cap_type: Port, length: 8, version: 1 }, capability_data: Port }, CapabilityMessage { header: CapabilityHeader { cap_type: Drive, length: 8, version: 2 }, capability_data: Drive }, CapabilityMessage { header: CapabilityHeader { cap_type: Smartcard, length: 8, version: 1 }, capability_data: Smartcard }], kind: ServerCoreCapabilityRequest })
2023-09-20T21:49:23.308548Z TRACE ironrdp_rdpdr: sending RdpdrPdu(CoreCapability { padding: 0, capabilities: [CapabilityMessage { header: CapabilityHeader { cap_type: General, length: 44, version: 2 }, capability_data: General(GeneralCapabilitySet { os_type: 0, os_version: 0, protocol_major_version: 1, protocol_minor_version: 12, io_code_1: IoCode1(RDPDR_IRP_MJ_CREATE | RDPDR_IRP_MJ_CLEANUP | RDPDR_IRP_MJ_CLOSE | RDPDR_IRP_MJ_READ | RDPDR_IRP_MJ_WRITE | RDPDR_IRP_MJ_FLUSH_BUFFERS | RDPDR_IRP_MJ_SHUTDOWN | RDPDR_IRP_MJ_DEVICE_CONTROL | RDPDR_IRP_MJ_QUERY_VOLUME_INFORMATION | RDPDR_IRP_MJ_SET_VOLUME_INFORMATION | RDPDR_IRP_MJ_QUERY_INFORMATION | RDPDR_IRP_MJ_SET_INFORMATION | RDPDR_IRP_MJ_DIRECTORY_CONTROL | RDPDR_IRP_MJ_LOCK_CONTROL), io_code_2: 0, extended_pdu: ExtendedPdu(RDPDR_DEVICE_REMOVE_PDUS | RDPDR_CLIENT_DISPLAY_NAME_PDU), extra_flags_1: ExtraFlags1(0x0), extra_flags_2: 0, special_type_device_cap: 1 }) }, CapabilityMessage { header: CapabilityHeader { cap_type: Smartcard, length: 8, version: 1 }, capability_data: Smartcard }], kind: ClientCoreCapabilityResponse })
2023-09-20T21:49:23.308569Z TRACE ironrdp_rdpdr: received RdpdrPdu(VersionAndIdPdu { version_major: 1, version_minor: 13, client_id: 3, kind: ServerClientIdConfirm })
2023-09-20T21:49:23.308572Z TRACE ironrdp_rdpdr: sending RdpdrPdu(ClientDeviceListAnnounce { device_list: [DeviceAnnounceHeader { device_type: Smartcard, device_id: 0, preferred_dos_name: PreferredDosName("SCARD"), device_data: [] }] })
2023-09-20T21:49:23.411110Z TRACE ironrdp_rdpdr: received RdpdrPdu::Unimplemented
2023-09-20T21:49:23.411115Z  WARN ironrdp_rdpdr: received unimplemented packet: RdpdrPdu::Unimplemented

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

2023-09-20T21:49:24.045425Z TRACE ironrdp_rdpdr: received RdpdrPdu(VersionAndIdPdu { version_major: 1, version_minor: 13, client_id: 2, kind: ServerAnnounceRequest })
2023-09-20T21:49:24.045439Z TRACE ironrdp_rdpdr: sending RdpdrPdu(VersionAndIdPdu { version_major: 1, version_minor: 12, client_id: 2, kind: ClientAnnounceReply })
2023-09-20T21:49:24.045447Z TRACE ironrdp_rdpdr: sending RdpdrPdu(Unicode("IronRDP"))
2023-09-20T21:49:24.133821Z TRACE ironrdp_rdpdr: received RdpdrPdu(CoreCapability { padding: 0, capabilities: [CapabilityMessage { header: CapabilityHeader { cap_type: General, length: 44, version: 2 }, capability_data: General(GeneralCapabilitySet { os_type: 2, os_version: 0, protocol_major_version: 1, protocol_minor_version: 13, io_code_1: IoCode1(RDPDR_IRP_MJ_CREATE | RDPDR_IRP_MJ_CLEANUP | RDPDR_IRP_MJ_CLOSE | RDPDR_IRP_MJ_READ | RDPDR_IRP_MJ_WRITE | RDPDR_IRP_MJ_FLUSH_BUFFERS | RDPDR_IRP_MJ_SHUTDOWN | RDPDR_IRP_MJ_DEVICE_CONTROL | RDPDR_IRP_MJ_QUERY_VOLUME_INFORMATION | RDPDR_IRP_MJ_SET_VOLUME_INFORMATION | RDPDR_IRP_MJ_QUERY_INFORMATION | RDPDR_IRP_MJ_SET_INFORMATION | RDPDR_IRP_MJ_DIRECTORY_CONTROL | RDPDR_IRP_MJ_LOCK_CONTROL | RDPDR_IRP_MJ_QUERY_SECURITY | RDPDR_IRP_MJ_SET_SECURITY), io_code_2: 0, extended_pdu: ExtendedPdu(RDPDR_DEVICE_REMOVE_PDUS | RDPDR_CLIENT_DISPLAY_NAME_PDU | RDPDR_USER_LOGGEDON_PDU), extra_flags_1: ExtraFlags1(0x0), extra_flags_2: 0, special_type_device_cap: 2 }) }, CapabilityMessage { header: CapabilityHeader { cap_type: Printer, length: 8, version: 1 }, capability_data: Printer }, CapabilityMessage { header: CapabilityHeader { cap_type: Port, length: 8, version: 1 }, capability_data: Port }, CapabilityMessage { header: CapabilityHeader { cap_type: Drive, length: 8, version: 2 }, capability_data: Drive }, CapabilityMessage { header: CapabilityHeader { cap_type: Smartcard, length: 8, version: 1 }, capability_data: Smartcard }], kind: ServerCoreCapabilityRequest })
2023-09-20T21:49:24.133851Z TRACE ironrdp_rdpdr: sending RdpdrPdu(CoreCapability { padding: 0, capabilities: [], kind: ClientCoreCapabilityResponse })
2023-09-20T21:49:24.133991Z TRACE ironrdp_rdpdr: received RdpdrPdu(VersionAndIdPdu { version_major: 1, version_minor: 13, client_id: 2, kind: ServerClientIdConfirm })
2023-09-20T21:49:24.133999Z TRACE ironrdp_rdpdr: sending RdpdrPdu(ClientDeviceListAnnounce { device_list: [] })
```

On the second time I'm sending back the empty `ClientDeviceListAnnounce` because it was already taken.

I'm not seeing this same behavior in Teleport. It may be being caused by the fact that we announce the smartcard capability but then currently don't handle one of the smartcard commands the server sends to us. I will continue development and then revisit whether this solves itself.

**Comment by CBenoit** [comment]
Oh okay, I see. Sounds good to me!

**Pull request merged**